### PR TITLE
[Collective][PR 3/6] Other collectives

### PR DESCRIPTION
**Pull request by zhisbug**
## Why are these changes needed?
This is the **third** PR for the project [Collective-in-Ray](https://github.com/ray-project/ray/pull/12174). 

To make each PR more manageable and friendly to reviewers, we break the entire project code into 6 incremental PRs:
See a list below:

1.  ~~(**merged**) Basic infrastructure; an in-actor collective interface `ray.util.collective.init_collective_group(*args, **kwargs)`; support for two collectives `allreduce` and `barrier`; some testing infrastructure, etc.~~
2. Driver-program interface, which includes: (1) the second interface: `actor.options(collective_options, ...).remote()` and the third interface `declare_collective_group(actors, collective_options, ...)`.
3. (**this one**) Support for other collectives: `allgather`, `broadcast`, `reduce`, `reducescatter`; refactor the tests into distributed tests and single-node cluster tests.
4. Communicator caching,  and support for num_gpus > 2 per actor/task.
5. CUDA stream management.
6. docs, examples, etc.

This is the third one (3/6).

## Related issue number
RFC #12174

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zhisbug** [comment]
@richardliaw

File: python/ray/util/collective/collective.py [hunk 1]
```diff
@@ -228,6 +228,111 @@ def barrier(group_name):
     g.barrier()
 
 
+def reduce(tensor,
+           dst_rank: int = 0,
+           group_name: str = "default",
+           op=types.ReduceOp.SUM):
+    """
+    Reduce the tensor across the group to the destination rank.
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:236]: for each of these methods, can you be sure to write the docstring in a way such that the first line has 1 sentence that does not have line-wrap? for ex:

```
"""Reduce the tensors across group to the destination rank.

Args:...
```

**Review by richardliaw — COMMENTED** [review]
Roughly looks good (skipped most of the test code!) 

Ping when comments addressed.

File: python/ray/util/collective/collective.py (hunk 1 above)
Comment (zhisbug) [reply python/ray/util/collective/collective.py:236]: done

**Comment by zhisbug** [comment]
@richardliaw all comments addressed. Test passed on my local cluster. Fine to merge!

**Review by richardliaw — APPROVED** [review]
thanks! will merge after master shows lint passes

**Pull request merged**